### ISSUE TITLE
Utilize Roslyn Source Generation to reduce boilerplate in (de)serialization

### ISSUE THREAD

**Issue by PaulusParssinen** [PaulusParssinen/Shockky#6]
The Shockwave file format includes many common patterns/structures and is _somewhat_ consistent with them. For example, arrays and some smaller common structures ones such as `Point`, `CastMemberId` and `Rect` are represented using common structure in binary. In addition to those, most "complex" and important one of them is internally called `VList` aka "Vector List". It's not actually "complex" but just annoying to write (de)serialization logic for and I have been thinking how to handle them best for few years now. `VList`s appear in many important resources of the file format, such as Cast member metadata (`VWCF`, later embedded to `CASt`) and Cross Cast Links (`ccl `) and more.

Now heavily inspired by [csnover's earthquake-rust `pvec!` macro](https://github.com/csnover/earthquake-rust/blob/wip/libearthquake/src/collections/pvec.rs) (seriously, this man's a genius!), I'm planning to implement something similar to make the handling `VList`s significantly easier by using build time Roslyn source generation.

This issue is just a tracking issue for my design for this model which is a work in progress.

## TODO: Design
Basically VLists consists of header and a offset table that points to entries in following it. 

I'm also currently in process of designing a source generation process for all the Lingo bytecode instructions which will share components with this but still essentially be seperate generator.

**Comment by PaulusParssinen** [comment]
Some source generation infrastructure is in place and currently used for Lingo instruction generation which is much simpler than generalized (de)serialization. This will essentially require developing a value type model which is constructed from Roslyn AST.

**Comment by PaulusParssinen** [comment]
Work started for this in https://github.com/PaulusParssinen/Shockky/commit/f331fc3df9c45ede36966f40f9a2f03dca18fb06

**Comment by emansom** [comment]
@Quackster 👀
